From 12639f0cd516b954cba599994987f2c3b8e7f155 Mon Sep 17 00:00:00 2001
From: Loren HUANG <b02279@freescale.com>
Date: Wed, 4 Sep 2013 15:58:04 +0800
Subject: [PATCH 0579/1072] ENGR00277333 gpu: Enable OT limitation for gc880

git://git.freescale.com/imx/linux-2.6-imx.git imx_3.10.17_1.0.0_beta
commit 561250ddcfbe1254ef783e4a0fc9b5e68e572708

Enable OT limitation for gc880, without this limitation
3D core may stall system bus when it is running at a very low clock.

Signed-off-by: Loren HUANG <b02279@freescale.com>
Acked-by: Shawn Guo
Signed-off-by: Hongbo Zhong <hongbo.zhong@windriver.com>
---
 .../arch/XAQ2/hal/kernel/gc_hal_kernel_hardware.c  |   15 +++++++++++++++
 1 files changed, 15 insertions(+), 0 deletions(-)

diff --git a/drivers/mxc/gpu-viv/arch/XAQ2/hal/kernel/gc_hal_kernel_hardware.c b/drivers/mxc/gpu-viv/arch/XAQ2/hal/kernel/gc_hal_kernel_hardware.c
index 00f3839..6492935 100644
--- a/drivers/mxc/gpu-viv/arch/XAQ2/hal/kernel/gc_hal_kernel_hardware.c
+++ b/drivers/mxc/gpu-viv/arch/XAQ2/hal/kernel/gc_hal_kernel_hardware.c
@@ -728,6 +728,13 @@ gckHARDWARE_Construct(
 
     default:
         hardware->type = gcvHARDWARE_3D;
+        if(hardware->identity.chipModel == gcv880)
+        {
+            /*set outstanding limit*/
+            gcmkONERROR(gckOS_ReadRegisterEx(Os, Core, 0x00414, &axi_ot));
+            axi_ot = (axi_ot & (~0xFF)) | 0x10;
+            gcmkONERROR(gckOS_WriteRegisterEx(Os, Core, 0x00414, axi_ot));
+        }
 
         if ((((((gctUINT32) (hardware->identity.chipFeatures)) >> (0 ? 9:9)) & ((gctUINT32) ((((1 ? 9:9) - (0 ? 9:9) + 1) == 32) ? ~0 : (~(~0 << ((1 ? 9:9) - (0 ? 9:9) + 1)))))) ))
         {
@@ -1198,6 +1205,14 @@ gckHARDWARE_InitializeHardware(
 #endif
 
     /* Limit 2D outstanding request. */
+    if(Hardware->identity.chipModel == gcv880)
+    {
+        gctUINT32 axi_ot;
+        gcmkONERROR(gckOS_ReadRegisterEx(Hardware->os, Hardware->core, 0x00414, &axi_ot));
+        axi_ot = (axi_ot & (~0xFF)) | 0x10;
+        gcmkONERROR(gckOS_WriteRegisterEx(Hardware->os, Hardware->core, 0x00414, axi_ot));
+    }
+
     if ((Hardware->identity.chipModel == gcv320)
         && ((Hardware->identity.chipRevision == 0x5007)
         || (Hardware->identity.chipRevision == 0x5220)))
-- 
1.7.5.4

